lib/AST/Decl.cpp

```diff
@@ -1824,44 +1847,53 @@ CanType ValueDecl::getOverloadSignatureType() const {
                            afd->getNumParameterLists())->getCanonicalType();
   }
 
-  if (isa<SubscriptDecl>(this)) {
-    CanType interfaceType = getInterfaceType()->getCanonicalType();
+  if (auto *asd = dyn_cast<AbstractStorageDecl>(this)) {
+    auto dc = getDeclContext();
 
-    // If the subscript declaration occurs within a generic extension context,
-    // consider the generic signature of the extension.
-    auto ext = dyn_cast<ExtensionDecl>(getDeclContext());
-    if (!ext) return interfaceType;
-
-    auto genericSig = ext->getGenericSignature();
-    if (!genericSig) return interfaceType;
-
-    if (auto funcTy = interfaceType->getAs<AnyFunctionType>()) {
-      return GenericFunctionType::get(genericSig,
-                                      funcTy->getParams(),
-                                      funcTy->getResult(),
-                                      funcTy->getExtInfo())
-              ->getCanonicalType();
+    // First, get the default overload signature type for the decl. For vars,
+    // this is the empty tuple type, as variables cannot be overloaded directly
+    // by type. For subscripts, it's their interface type.
+    CanType defaultSignatureType;
+    if (isa<VarDecl>(this)) {
+      defaultSignatureType = TupleType::getEmpty(getASTContext());
+    } else {
+      defaultSignatureType = getInterfaceType()->getCanonicalType();
     }
 
-    return interfaceType;
+    // If we're not in a type context, we're done.
+    // Otherwise, we want to curry the default signature type with the 'self'
+    // type of the given context in order to ensure the overload signature type
+    // is unique across different contexts, such as between a protocol extension
+    // and struct decl.
+    if (!dc->isTypeContext())
+      return defaultSignatureType;
+
+    // Treat the declared interface type of the context as the 'self' type to
+    // curry the signature with. Additionally, if the storage decl is static,
+    // work with a metatype.
+    auto selfTy = dc->getDeclaredInterfaceType();
+    if (asd->isStatic())
+      selfTy = MetatypeType::get(selfTy);
+
+    auto parentDecl = dc->getAsDeclOrDeclExtensionContext();
+    auto genericCtx = parentDecl->getAsGenericContext();
+
+    // If we're in a context with a generic signature, return a generic function
+    // type with that signature. Otherwise, return a non-generic type.
```

Comment:
What if the subscript itself is generic? I know that didn't work before either, but there should at least be a FIXME if you're going to redo the code without fixing it.

Reply:
@jrose-apple Looking into it, I think we already correctly handle generic subscripts in this PR. The generic signature just gets nested in the type when we curry it with 'self'.

For example, for this subscript:

```swift
struct Foo<U> {
  subscript<T>(_ t: T) -> U {
    fatalError()
  }
}
```

Its overload type is:
```
(generic_function_type escaping
  (input=paren_type
    (bound_generic_struct_type decl=main.(file).Foo
      (generic_type_param_type depth=0 index=0)))
  (output=generic_function_type escaping
    (input=paren_type
      (generic_type_param_type depth=1 index=0))
    (output=generic_type_param_type depth=0 index=0)
    ( generic_sig=<τ_0_0, τ_1_0>))
  ( generic_sig=<τ_0_0>))
```

I think this is okay for the purposes for an overload type. I'll go ahead and add some tests for generic subscripts though.

Reply:
I was pretty sure it still has to be a GenericFunctionType, though, which is something that *wouldn't* happen if the subscript is generic but the context is not.

(I'll reply to the other comments later.)

Reply:
@jrose-apple Ah okay, that makes sense. So presumably we need to lift the subscript's generic signature up to the curried function's generic signature.

Actually [just found an existing function](https://github.com/apple/swift/blob/d8b355dbaebd157ba75e19fbc6ace87ef88e3e98/lib/Sema/CSRanking.cpp#L257) that curries a type with a given context's 'self', so can just reuse that for this (it doesn't use a MetatypeType for a static member, but we didn't really need that to begin with as static members don't have conflicting signatures with instance members).

So for this subscript:
```swift
struct Foo {
  subscript<T>(_ t: T) -> T {
    fatalError()
  }
}
```
its overload type is now:
```
(generic_function_type escaping
  (input=paren_type
    (struct_type decl=main.(file).Foo))
  (output=function_type escaping
    (input=paren_type
      (generic_type_param_type depth=0 index=0))
    (output=generic_type_param_type depth=0 index=0))
  ( generic_sig=<τ_0_0>))
```

which matches up with the overload type for a similar method decl, e.g:
```swift
struct Foo {
  func foo<T>(_ t: T) -> T {
    fatalError()
  }
}
```

I've gone ahead and pushed the revisions for your feedback – thanks for your patience!